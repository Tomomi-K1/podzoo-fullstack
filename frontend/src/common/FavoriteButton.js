import React, {useState, useContext} from "react";
import UserContext from "../UserContext";
// Material UI
import FormControlLabel from '@mui/material/FormControlLabel';
import Checkbox from "@mui/material/Checkbox";
import FavoriteBorder from "@mui/icons-material/FavoriteBorder";
import Favorite from "@mui/icons-material/Favorite";
const label = { inputProps: { "aria-label": "like button" }};

/** FavoriteButton
 * shows favorite button
 * makes api call to backend to remove and add user's favorite
 * update heart colored when it is clicked
 * component using FavoriteButton : PodcastDetailedLayout.js
 */
function FavoriteButton({podcastData, feedId}) {  
    const { currentUser, likePod, removeLike, checkFavPod } = useContext(UserContext);
    const [checked, setChecked] = useState(()=> checkFavPod(feedId));
    const username = currentUser.username;
    // make podcastData suitable for backend database
    const data = {
        feedId: podcastData.id,
        author: podcastData.author,
        title: podcastData.title,
        artwork: podcastData.artwork
    }

     /** == handleChange function  ==
      * Update checkbox information field
      * runs removeLike, LikePod functions
     */
    function handleChange (evt){        
        if(evt.target.checked === true){
            setChecked(true);
            likePod(username, feedId, data);
        } else{
            setChecked(false);
            removeLike(username, feedId);    
        }
      };
   
  return (
    <form>
        <FormControlLabel 
            control={
                    <Checkbox
                    {...label}
                    checked={checked}
                    color='error'
                    icon={<FavoriteBorder />}
                    checkedIcon={<Favorite />}
                    onChange={handleChange} />
                    
                    }
            label ="Favorite" />
  </form>
  );
}

export default FavoriteButton;