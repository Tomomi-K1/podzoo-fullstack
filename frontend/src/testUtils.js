import React from "react";
import UserContext from "./UserContext";

const demoUser = {
  username: "testuser",
  email: "test@test.net",
  fav:[1,2,3]
};

const favoritesArray =[1, 2]

const UserProvider =
    ({ children, currentUser = demoUser, checkFavPod=()=>true, favorites= favoritesArray}) => (
    <UserContext.Provider value={{ currentUser, checkFavPod, favorites }}>
      {children}
    </UserContext.Provider>
);

export { UserProvider };