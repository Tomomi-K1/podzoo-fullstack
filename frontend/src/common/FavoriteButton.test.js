import React from "react";
import { render, screen } from "@testing-library/react";
import { MemoryRouter } from "react-router";
import FavoriteButton from "./FavoriteButton";
import { UserProvider } from "../testUtils";

const data = {  
    feedId: 1,
    author: 'test-author',
    title: 'test-title',
    artwork: 'test-url'};
const feedId = '1'

it('renders', () => {
    render(
        <MemoryRouter>
            <UserProvider>
                <FavoriteButton podcastData={data} feedId={feedId} />
            </UserProvider>
        </MemoryRouter>
    ) 
})

it('matches snapshot', () => {
    const { asFragment } = render(
        <MemoryRouter>
            <UserProvider>
            <FavoriteButton podcastData={data} feedId={feedId} />
            </UserProvider>
        </MemoryRouter>
    )
    expect(asFragment()).toMatchSnapshot(); 
})
