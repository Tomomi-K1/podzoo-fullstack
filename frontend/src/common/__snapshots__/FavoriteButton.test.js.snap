// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<DocumentFragment>
  <form>
    <label
      class="MuiFormControlLabel-root MuiFormControlLabel-labelPlacementEnd css-j204z7-MuiFormControlLabel-root"
    >
      <span
        class="MuiButtonBase-root MuiCheckbox-root MuiCheckbox-colorError PrivateSwitchBase-root MuiCheckbox-root MuiCheckbox-colorError Mui-checked MuiCheckbox-root MuiCheckbox-colorError css-1peyhh5-MuiButtonBase-root-MuiCheckbox-root"
      >
        <input
          aria-label="like button"
          checked=""
          class="PrivateSwitchBase-input css-1m9pwf3"
          data-indeterminate="false"
          type="checkbox"
        />
        <svg
          aria-hidden="true"
          class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
          data-testid="FavoriteIcon"
          focusable="false"
          viewBox="0 0 24 24"
        >
          <path
            d="m12 21.35-1.45-1.32C5.4 15.36 2 12.28 2 8.5 2 5.42 4.42 3 7.5 3c1.74 0 3.41.81 4.5 2.09C13.09 3.81 14.76 3 16.5 3 19.58 3 22 5.42 22 8.5c0 3.78-3.4 6.86-8.55 11.54L12 21.35z"
          />
        </svg>
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </span>
      <span
        class="MuiTypography-root MuiTypography-body1 MuiFormControlLabel-label css-ahj2mt-MuiTypography-root"
      >
        Favorite
      </span>
    </label>
  </form>
</DocumentFragment>
`;
